### PR TITLE
feature/multi-message-command

### PR DESCRIPTION
**Pull request by ferrom-jaia**
Video found here:
https://jaia-innovation.atlassian.net/wiki/spaces/PD/pages/875888645/Multi-message+command+2022-08-12

### CODE REVIEW

File: src/lib/messages/jaia_dccl.proto [hunk 1]
```diff
@@ -13,7 +13,7 @@ message Command
 {
     option (dccl.msg) = {
         id: 80
-        max_bytes: 200
+        max_bytes: 552
```
Comment (tsaubergine) [inline src/lib/messages/jaia_dccl.proto]: We can't arbitrarily make this larger than the MTU supported by the XBee driver.

This needs to stay 200, and max_repeat for goal 14.

You can set more than 14 goals on hub when we're encoding using Protobufs.

File: src/bin/hub_manager/hub_manager.cpp [hunk 2]
```diff
@@ -271,3 +275,141 @@ void jaiabot::apps::HubManager::handle_dive_packet(const jaiabot::protobuf::Dive
     // republish
     interprocess().publish<jaiabot::groups::dive_packet>(dive_packet);
 }
+
+void jaiabot::apps::HubManager::handle_command(const jaiabot::protobuf::Command& input_command)
+{
+    using protobuf::Command;
+    auto command = input_command;
+    std::vector<Command> command_fragments;
+    int goal_max_size = 14;
```
Comment (tsaubergine) [inline src/bin/hub_manager/hub_manager.cpp]: would love to see this pulled out of the protobuf message:

Should be:
```
Command::descriptor()
                                   ->FindFieldByName("goal")
                                   ->options()
                                   .GetExtension(dccl::field)
                                   .max_repeat();
```

File: src/bin/hub_manager/hub_manager.cpp [hunk 3]
```diff
@@ -271,3 +275,141 @@ void jaiabot::apps::HubManager::handle_dive_packet(const jaiabot::protobuf::Dive
...
+                    if (command.plan().goal(goal_index).has_name())
+                    {
+                        goal->set_name(command.plan().goal(goal_index).name());
+                    }
+                    if (command.plan().goal(goal_index).has_task())
+                    {
+                        protobuf::MissionTask* task = new protobuf::MissionTask();
+                        task->ParseFromString(
+                            command.plan().goal(goal_index).task().SerializeAsString());
+                        goal->set_allocated_task(task);
```
Comment (tsaubergine) [inline src/bin/hub_manager/hub_manager.cpp]: What the??

Just copy:
```
*goal->mutable_task() = command.plan().goal(goal_index).task();
```

File: src/bin/hub_manager/hub_manager.cpp [hunk 4]
```diff
@@ -271,3 +275,141 @@ void jaiabot::apps::HubManager::handle_dive_packet(const jaiabot::protobuf::Dive
...
+                    }
+
+                    protobuf::GeographicCoordinate* coord = new protobuf::GeographicCoordinate();
+                    coord->ParseFromString(
+                        command.plan().goal(goal_index).location().SerializeAsString());
+                    goal->set_allocated_location(coord);
```
Comment (tsaubergine) [inline src/bin/hub_manager/hub_manager.cpp]: See above, no need for SerialToString and then Parse. Very inefficient and confusing:

```
*goal->mutable_location() = command.plan.goal(goal_index).location();
```

File: src/bin/mission_manager/app.cpp [hunk 5]
```diff
@@ -366,3 +380,110 @@ void jaiabot::apps::MissionManager::handle_command(const protobuf::Command& comm
...
+            if (initial_fragment.plan().has_movement())
+            {
+                out_command.mutable_plan()->set_movement(initial_fragment.plan().movement());
+            }
+
+            if (initial_fragment.plan().has_recovery())
+            {
+                protobuf::MissionPlan_Recovery* recovery = new protobuf::MissionPlan_Recovery();
+                recovery->ParseFromString(initial_fragment.plan().recovery().SerializeAsString());
+                out_command.mutable_plan()->set_allocated_recovery(recovery);
```
Comment (tsaubergine) [inline src/bin/mission_manager/app.cpp]: See above.

File: src/bin/mission_manager/app.cpp [hunk 6]
```diff
@@ -366,3 +380,110 @@ void jaiabot::apps::MissionManager::handle_command(const protobuf::Command& comm
...
+                        protobuf::MissionTask* task = new protobuf::MissionTask();
+                        task->ParseFromString(
+                            fragment.second.plan().goal(goal_index).task().SerializeAsString());
+                        goal->set_allocated_task(task);
+                    }
+
+                    protobuf::GeographicCoordinate* coord = new protobuf::GeographicCoordinate();
+                    coord->ParseFromString(
+                        fragment.second.plan().goal(goal_index).location().SerializeAsString());
+                    goal->set_allocated_location(coord);
```
Comment (tsaubergine) [inline src/bin/mission_manager/app.cpp]: See above

File: src/bin/mission_manager/app.cpp [hunk 7]
```diff
@@ -366,3 +380,110 @@ void jaiabot::apps::MissionManager::handle_command(const protobuf::Command& comm
...
+                    if (fragment.second.plan().goal(goal_index).has_name())
+                    {
+                        goal->set_name(fragment.second.plan().goal(goal_index).name());
+                    }
+                    if (fragment.second.plan().goal(goal_index).has_task())
+                    {
+                        protobuf::MissionTask* task = new protobuf::MissionTask();
+                        task->ParseFromString(
+                            fragment.second.plan().goal(goal_index).task().SerializeAsString());
+                        goal->set_allocated_task(task);
```
Comment (tsaubergine) [inline src/bin/mission_manager/app.cpp]: See above

File: src/lib/messages/mission.proto [hunk 8]
```diff
@@ -178,6 +178,8 @@ message MissionPlan
         ];
     }
     optional Speeds speeds = 5;
+    optional uint32 fragment_index = 6 [(dccl.field) = { min: 0 max: 255 }];
```
Comment (tsaubergine) [inline src/lib/messages/mission.proto]: Isn't this hardcoded to max: 2 at this point (for three fragments)? Why 255 then?

File: src/lib/messages/mission.proto [hunk 9]
```diff
@@ -178,6 +178,8 @@ message MissionPlan
...
+    optional uint32 expected_fragments = 7 [(dccl.field) = { min: 0 max: 255 }];
```
Comment (tsaubergine) [inline src/lib/messages/mission.proto]: Same as above, should be 1->3 presumably

File: src/bin/hub_manager/hub_manager.cpp [hunk 10]
```diff
@@ -271,3 +275,141 @@ void jaiabot::apps::HubManager::handle_dive_packet(const jaiabot::protobuf::Dive
...
+            if (command.plan().has_movement() && fragment_index == 0)
+            {
+                command_fragment.mutable_plan()->set_movement(command.plan().movement());
+            }
+
+            if (command.plan().has_recovery() && fragment_index == 0)
+            {
+                protobuf::MissionPlan_Recovery* recovery = new protobuf::MissionPlan_Recovery();
+                recovery->ParseFromString(command.plan().recovery().SerializeAsString());
+                command_fragment.mutable_plan()->set_allocated_recovery(recovery);
```
Comment (tsaubergine) [inline src/bin/hub_manager/hub_manager.cpp]: Same as elsewhere, just copy (one line).

**Review by tsaubergine — CHANGES_REQUESTED** [review]

File: src/bin/hub_manager/hub_manager.cpp (hunk 4 above)
Comment (ferrom-jaia) [reply src/bin/hub_manager/hub_manager.cpp]: Ya, this is way better!

File: src/lib/messages/mission.proto (hunk 8 above)
Comment (ferrom-jaia) [reply src/lib/messages/mission.proto]: Good point

**Review by tsaubergine — DISMISSED** [review]

**Review by jason-jaia — DISMISSED** [review]
Code review looks swell.

**Pull request merged**